# Solución Problema Optimizacion - SymPy
Empezamos importando las librerias

In [ ]:
from sympy import Matrix, symbols, solve
x, y = symbols('x, y')

$f(x)=x_{1}^{4}+x_{1}^{2}\left(1-2 x_{2}\right)+2 x_{2}^{2}-2 x_{1} x_{2}+4.5 x_{1}-4 x_{2}+4$

In [ ]:
f = x**4 + x**2*(1-2*y) + 2*y**2 - 2*x*y + 4.5*x - 4*y + 4

Calculamos el gradiente

In [ ]:
Df = Matrix([f]).jacobian(Matrix(list(f.free_symbols)))
Df

Matrix([[4*x**3 + 2*x*(1 - 2*y) - 2*y + 4.5, -2*x**2 - 2*x + 4*y - 4]])

Encontramos la solucion para $∇f = \mathbf{0}$

In [ ]:
estaf = solve(Df, [x, y], dict=True)
estaf

[{x: -1.05274130826315 + 2.71050543121376e-20*I,
  y: 1.02776147693023 + 0.e-23*I},
 {x: 0.611731035834618 - 1.35525271560688e-20*I,
  y: 1.49297294801896 + 0.e-20*I},
 {x: 1.94101027242853 - 2.71050543121376e-20*I,
  y: 3.85426557505081 - 0.e-22*I}]

Calculamos la Hessiana

In [ ]:
Df2 = Df.jacobian(Matrix(list(f.free_symbols)))
Df2

Matrix([
[12*x**2 - 4*y + 2, -4*x - 2],
[         -4*x - 2,        4]])

Sustituimos el valor en cada uno de los puntos estacionarios que se encontraron previamente. Para cada uno obtenemos la matriz Hessiana y calculamos los valores propios.

In [ ]:
Df2.subs(estaf[0]).eigenvals()

{11.813737176262 - 6.90970427757896e-19*I: 1,
 3.37438806150044 + 6.03390054900417e-21*I: 1}

El primer punto es un minimo, debido que sus valores propios son positivos.

In [ ]:
Df2.subs(estaf[1]).eigenvals()

{-2.51611694353881 - 1.03760499496677e-19*I: 1,
 7.03480347390251 - 1.2231659022407e-19*I: 1}

El segundo punto es un punto de ensilladura, un valor propio es positivo y el otro es negativo.

In [ ]:
Df2.subs(estaf[2]).eigenvals()

{34.8804637110769 - 1.20869731539304e-18*I: 1,
 0.912724520796989 - 5.22771512226742e-20*I: 1}

El tercer punto es un minimo, los valores propios de la matriz Hessiana son positivos.

Ahora analizamos para el primer y segundo punto estacionario cual es el valor de la funcion objetivo

In [ ]:
f.subs(estaf[0])

-4.84838179490511 + (-1.05274130826315 + 2.71050543121376e-20*I)**4 - 2*(-1.05274130826315 + 2.71050543121376e-20*I)*(1.02776147693023 + 0.e-23*I) + 2*(1.02776147693023 + 0.e-23*I)**2 + (-1.05552295386046 - 0.e-22*I)*(-1.05274130826315 + 2.71050543121376e-20*I)**2 + 1.21866865286212e-19*I

Es dificil obtener un resultado directamente con SymPy, por eso volvemos a crear la funcion de la funcion objetivo

In [ ]:
fobj = lambda x,y: x**4 + x**2*(1-2*y) + 2*y**2 - 2*x*y + 4.5*x - 4*y + 4

Evaluamos la funcion obejtivo en el primer punto estacionario

In [ ]:
fobj(estaf[0][x].as_real_imag()[0],estaf[0][x].as_real_imag()[1])

1.59917804963983

Evaluamos la funcion objetivo en el segundo punto estacionario

In [ ]:
fobj(estaf[2][x].as_real_imag()[0],estaf[2][x].as_real_imag()[1])

30.6962806673041

El minimo global corresponde al primer punto estacionario (-1.53, 1.03) con un valor de la funcion objeetivo igual a 1.6.